### ISSUE TITLE
Find a way of getting data on when players were subbed on/off, and when goals were scored

### ISSUE THREAD

**Issue by nbarlowATI** [alan-turing-institute/AIrsenal#430]
Currently from the FPL API we have no way of knowing:
* At what minute each goal was scored
* If a player plays < 90 mins, did they start and get subbed off, or get subbed on?  (Or get subbed on and then subbed off, in a few cases!)

This means that we don't know whether a player was on the pitch when a goal was scored, which would be useful information for evaluating their defensive strength.

Is there some other API we could get this from?  (maybe https://www.football-data.org/ or similar?)
Or scrape from match reports?

**Comment by chahak13** [comment]
Hi, I have been reading up a bit on this project and would like to contribute more to this project. I was thinking maybe I can help collect this data? I also checked that football-data API has the minute information for goals scored and the substitution information too. I can try and integrate both these things if that's okay?

**Comment by jack89roberts** [comment]
Sure, that would be great @chahak13 ! Let us know if there's anything we can help with.

**Comment by chahak13** [comment]
Hi, I wanted your opinion on where exactly do we want to store the goal-scoring time data? In the player details JSON or in the fixture information CSV? I guess since the goal data is being stored for individual players which is then being summed, this info would make sense to be put in the individual player info itself but I just wanted to confirm it once.

**Comment by jack89roberts** [comment]
Right, good question. I think we need two things (which can be treated and saved separately):

- For every goal in every fixture, which team scored it and what minute it was scored (and perhaps which player scored it, we may not strictly need this but if it's available without much extra work we may as well include it).
- For every player in every fixture, what minute they started and what minute they were subbed off (or 0 and 90 if they played the whole match).

As a starting point I think it would be fine (and maybe preferable actually) for these to be two separate new files (per season, and back to the 15/16 season if possible eventually).

Ultimately we'll need to decide where to add these in the database. Maybe:
- A new `Goal` table with fields `fixture`, `minute`, `team`, `player`.
- `minute_started` and `minute_finished` fields added to the `PlayerScore` table.

Also no need to do all of this by the way (though you're welcome to!), e.g.  a pull request with working API queries would be a great start.

**Comment by chahak13** [comment]
Okay, got it. Thanks! I'll start working on getting the goal scored time first. There is one small issue though, that I did not foresee. The football-data.org API doesn't have this information in the free tier but in the free+three tier which is 25 EUR/month. If we do have that subscription, then we can get the time scored, the player who scored, and substitutions information in the same response. Can you please suggest how should I proceed?

**Comment by jack89roberts** [comment]
Ah right, that's a shame. It's not something we have a subscription for (and we want AIrsenal to be as accessible as possible). Really we'd want something that's free and we can use the data as we like. I'll try to have a search around to see if there are any other sources that look promising.

**Comment by jack89roberts** [comment]
Understat might be worth a try, I'm not sure they have an API but people have got data from there if you search around: https://understat.com/ (usually for expected goal stats but they might have other info available)

**Comment by chahak13** [comment]
That's great, thanks! I'll look at that and also see if I can find something else.

**Comment by jack89roberts** [comment]
If you do look at Understat I know this repo has some data available: https://github.com/vaastav/Fantasy-Premier-League (not what we need for this but may be a helpful start, e.g. [this script](https://github.com/vaastav/Fantasy-Premier-League/blob/master/understat.py) and [an example file](https://github.com/vaastav/Fantasy-Premier-League/blob/master/data/2021-22/understat/Adam_Armstrong_4419.csv)).

Time of subs and goals is available on the web page for each match (e.g. https://understat.com/match/16435) but probably not in an easy to read/parse format.

**Comment by chahak13** [comment]
I did take a look at the repo you mentioned. I believe that's also the source for the data files that we're currently looking at? There's also another [understat python library](https://github.com/amosbastian/understat) which provides an interface to understat but it doesn't seem to be pulling in goal information. Scraping understat was also what I was going to try next.  

I did find [this](https://www.api-football.com/documentation-v3#operation/get-fixtures-events) API which has the event data that we want but seems to be providing 100 requests/day in the free plan. I'm not sure if that will be enough for us to repopulate all the previous data but maybe for new matchdays it might be sufficient? What are your thoughts about it?

PS: Sorry but I'll not be able to work on this much for the next two days since I have a couple of assignments going on right now.

**Comment by jack89roberts** [comment]
No rush in getting this done, good luck with the assignments!

API-football looks interesting, I think I've come across it before but haven't used it. 100 requests/day is probably enough to do something useful (assuming we only need one request per match), we could set up a script to slowly gather the data for past seasons (would take about 4 days per season). We should check they'd allow us to save data from there in the repo,  though (I can do that once we confirm the API does have what we need).

**Comment by jack89roberts** [comment]
Looking at the docs it might be possible to request more than one fixture at a time. If that's the case the 100 requests would go a long way.

**Comment by chahak13** [comment]
Hey @jack89roberts, I decided on scraping the understat site to get the information we need. I've used BeautifulSoup to do it right now, it's slightly slow but I still got all the info for EPL 2020 in 4 mins so it's not that bad either. I've attached the information as a JSON file. Let me know your thoughts about it. 
[epl_2020.txt](https://github.com/alan-turing-institute/AIrsenal/files/7318932/epl_2020.txt)

Example:
```json
{
    match_id: {
        "home": home_team,
        "away": away_team,
        "goals": [
            [
                goal_scorer,
                goal_time
            ]
        ],
        "subs": [
            [
                player_out,
                player_in,
                sub_time
            ]
        ]
    }
}
```

```json
{
    "14086": {
        "home": "Fulham",
        "away": "Arsenal",
        "goals": [
            [
                "Alexandre Lacazette",
                "8"
            ],
            [
                "Gabriel",
                "48"
            ],
            [
                "Pierre-Emerick Aubameyang",
                "56"
            ]
        ],
        "subs": [
            [
                "Neeskens Kebano",
                "Franck Zambo",
                "64"
            ],
            [
                "Josh Onomah",
                "Bobby Reid",
                "76"
            ],
            [
                "Willian",
                "Nicolas Pepe",
                "77"
            ],
            [
                "Granit Xhaka",
                "Dani Ceballos",
                "80"
            ],
            [
                "Alexandre Lacazette",
                "Eddie Nketiah",
                "88"
            ]
        ]
    },
    "14087": {
        "home": "Crystal Palace",
        "away": "Southampton",
        "goals": [
            [
                "Wilfried Zaha",
                "12"
            ]
        ],
        "subs": [
            [
                "Jan Bednarek",
                "Jannik Vestergaard",
                "48"
            ],
            [
                "James McCarthy",
                "Luka Milivojevic",
                "76"
            ],
            [
                "William Smallbone",
                "Moussa Djenepo",
                "79"
            ],
            [
                "Jeffrey Schlupp",
                "Eberechi Eze",
                "83"
            ],
            [
                "Che Adams",
                "Shane Long",
                "87"
            ]
        ]
    },
    "14090": {
        "home": "Liverpool",
        "away": "Leeds",
        "goals": [
            [
                "Jack Harrison",
                "11"
            ],
            [
                "Virgil van Dijk",
                "19"
            ],
            [
                "Patrick Bamford",
                "29"
            ],
            [
                "Mohamed Salah",
                "32"
            ],
            [
                "Mateusz Klich",
                "65"
            ]
        ],
        "subs": [
            [
                "Naby Keita",
                "Fabinho",
                "60"
            ],
            [
                "Patrick Bamford",
                "Rodrigo",
                "64"
            ],
            [
                "Jordan Henderson",
                "Curtis Jones",
                "68"
            ],
            [
                "Mateusz Klich",
                "Jamie Shackleton",
                "83"
            ],
            [
                "Trent Alexander-Arnold",
                "Joel Matip",
                "91"
            ]
        ]
    },
  ```

**Comment by jack89roberts** [comment]
Nice looks good, definitely would be happy to have a pull request for this! Couple of small comments (but they don't have to be done for the PR):
- If it's possible to include a date/time for each match that would be nice. It's not strictly necessary (so only add it if it's straightforward), but that would allow an extra check when linking to data from other sources.
- Similarly, not strictly necessary but including the team the goal/sub relates to would also be helpful for sanity checks.
- I noticed the last sub in the example above is at 91 minutes (i.e. in stoppage time). How does this work at the end of the first half? Could we distinguish between 45+2 mins (stoppage time) and 47 mins (2 mins into 2nd half)?
- 4 mins to get data for past seasons is fine I think, but can it be quicker for getting data for the current season? E.g. if we just need the data for the last gameweek (10 matches) is there a way to query only those and not the whole season?

I also noticed there's a Understat Python package: https://github.com/amosbastian/understat . I don't think they have this data available so you could see if they'd be interested having this contributed over there too (but definitely do put it in an AIrsenal pull request 😄 )
